src/Microsoft.Framework.Runtime/Internal/ConcurrencyUtilities.cs

```diff
@@ -29,25 +29,35 @@ public static void ExecuteWithFileLocked(string filePath, Action<bool> action)
             .GetAwaiter().GetResult();
         }
 
-        public async static Task<T> ExecuteWithFileLocked<T>(string filePath, Func<bool, Task<T>> action)
+        public async static Task<T> ExecuteWithFileLocked<T>(string filePath, int timeoutMilliseconds, Func<bool, Task<T>> action)
         {
-            var createdNew = false;
-            var fileLock = new Semaphore(initialCount: 0, maximumCount: 1, name: FilePathToLockName(filePath),
-                createdNew: out createdNew);
-            try
+            for (var i = 0; i < 3; ++i)
             {
-                // If this lock is already acquired by another process, wait until we can acquire it
-                if (!createdNew)
+                var createdNew = false;
+                var fileLock = new Semaphore(initialCount: 0, maximumCount: 1, name: FilePathToLockName(filePath),
+                    createdNew: out createdNew);
+                try
                 {
-                    fileLock.WaitOne();
-                }
+                    // If this lock is already acquired by another process, wait until we can acquire it
+                    if (!createdNew)
+                    {
+                        var signaled = fileLock.WaitOne(5000);
```

Comment:
You're not using `timeoutMilliseconds` I would also just change it to be a TimeSpan